### ISSUE TITLE
Simplifying arrays of tables

### ISSUE THREAD

**Issue by jasonhansel** [toml-lang/toml#327]
I have two ideas for simplifying how arrays of tables work in TOML. The first one is motivated by the following example:

```
[[fruit]]
type = "apple"
[fruit.physical]
color = "red"
```

In this case it is clear that "fruit," after the "[[fruit]]" entry, refers to the last item added to the "fruit" array. However, when one writes:

```
[[fruit]]
type = "apple"
[fruit]
color = "red"
```

It is, if I understand the spec correctly, a syntax error. I think that it would be simpler if this TOML code produced the following JSON:

```
{ fruit: [ { type: "apple", color: "red" } ] }
```

My second idea relates to the automatic creation of nested tables. Currently, when you use `[config.database.mysql]`, for instance, the tables `config.database` and `config` are auto-created. However, consider this case:

```
[[fruit]]
type = "apple"
[fruit.physical]
color = "red"
```

This outputs the JSON cod:

```
{ fruit:  [ {type: "apple", physical: {color: "red" } } ] } 
```

But what if one wanted to create a new fruit and add a "physical" key to that, instead? One would need to write:
```
[[fruit]]
type = "apple"
[[fruit]] # awkward
[fruit.physical]
color = "red"
```

I propose that instead we allow the following syntax

```
[[fruit]]
type = "apple"
[[fruit].physical]  # Note closing brace position
color = "red"
```

Equivalent to the JSON code: 

```
{ fruit:  [ {type: "apple" } ], [ { physical: {color: "red" } } ] } 
```

In short, this would enable the automatic creation of nested tables within arrays.

Do either of these sound like good ideas? Could either of them be incorporated into a future version of the spec? Thanks!

**Comment by BurntSushi** [comment]
> ```
> [[fruit]]
> type = "apple"
> [fruit]
> color = "red"
> ```
>
> It is, if I understand the spec correctly, a syntax error. I think that it would be simpler if this TOML code produced the following JSON:
>
> ```
> { fruit: [ { type: "apple", color: "red" } ] }
> ```

This doesn't make any sense to me. Why not just use:

```toml
[[fruit]]
type = "apple"
color = "red"
```

This should produce the same JSON.


> But what if one wanted to create a new fruit and add a "physical" key to that, instead? One would need to write:
> 
> ```
> [[fruit]]
> type = "apple"
> [[fruit]] # awkward
> [fruit.physical]
> color = "red"
> ```

What is awkward about it? It looks simple and explicit to me.

**Comment by jakwings** [comment]
Both proposals don't look good to me too. We should write in a simple way and let the configuration be easy to parse by human eyes.

Switching between `[[fruit]]` and `[[fruit].physical]` is not a good thing.

If you want to make it beautiful, just use more indent. Like this:

```
[[fruit]]
type = "apple"
  [fruit.physical]
  color = "red"

[[fruit]]
type = "apple"

[[fruit]] # good
  [fruit.physical]
  color = "red"
```

You can have a look at my encoder online: https://jakwings.github.io/tomlify-j0.4/demo/

The indented TOML text can be seen in the debug console.

**Comment by jasonhansel** [comment]
These are all good points. This issue should probably be closed.
